Create rule S6907: Environment variables names for a container should not be duplicated (#3606)
SonarSource/rspec: rules/S6907/kubernetes/rule.adoc

Having duplicate environment variable names for a container can lead to unexpected behavior, as the last declared variable will overwrite any previous ones with the same name. This rule aims to detect and prevent such situations, thereby enhancing the clarity and predictability of Kubernetes configurations.

== Why is this an issue?

If an environment variable of a container is declared multiple times with different values, it can become unclear which value will be used by the application and can lead to unpredictable behavior. Also, if an issue arises related to the environment variable, debugging can be challenging. The problem could be due to the overwritten variable, but this might take time to be noticeable.

== How to fix it

> Identify the duplicate environment variables for a container and remove the duplicates manually. If the same variable is needed with different values, consider using a different name for each variable. An environment variable for a container is defined in `env[*].name` directly by providing a value or in `envFrom` by referencing either a `ConfigMap` or a `Secret`

Reviewer: ```suggestion
Identify the duplicate environment variables for a container and remove the duplicates manually. If the same variable is needed with different values, consider using a different name for each variable.
```

=== Code examples

==== Noncompliant code example

[source,text,diff-id=1,diff-type=noncompliant]
----
apiVersion: v1
kind: Pod
metadata:
  name: env-example
spec:
  containers:
  - name: nginx
    image: nginx:1.7.9
    env:
    - name: VAR1
      value: "value1"
    - name: VAR1 # Noncompliant

>       value: "value2"

Reviewer: We should raise with secondary locations.

----

==== Compliant solution

[source,text,diff-id=1,diff-type=compliant]
----
apiVersion: v1
kind: Pod
metadata:
  name: env-example
spec:
  containers:
  - name: nginx
    image: nginx:1.7.9
    env:
    - name: VAR1
      value: "value1"
    - name: VAR2
      value: "value2"
----

> `ConfigMap` and `Secret` may also be defined in a different file where this rule has to be applied too.

Reviewer: Sounds a bit too vague. Should this already be included in the rule and can be implemented meaningfully without x-file analysis?
yes -> include it with a proper example.
no -> Remove it.

Reviewer 2: This is similar to what we did for LimitRange, we are mentioning it here but the example should be added along with the split ticket

Reviewer: Yes but still the sentence sounds too vague. If a user doesn't know of configMap or secret he wouldn't understand it.
For now I would not include this, as it would not fix the already existing issue but provide the user with more detective work looking in ConfigMap or Secret.

== Resources
=== Documentation
* https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/[Kubernetes - Environment Variables for a Container]

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

=== Message

> Resolve the duplicate name for the environment variable.

Reviewer: ```suggestion
Resolve the duplication of this environment variable.
```

=== Highlighting

> * Highlight all the location where duplicated environment variable is being used.

Reviewer: ```suggestion
* Highlight all the locations where duplicated environment variables are being used.
```

> * Highlight the variable name if already declared and assigned a value before. In case of duplicate keys for ConfigMap or Secret, highlight the key name which is used last.

Reviewer: We should highlight all locations of the variable name.
I don't quite understand this part of the sentence: `if already declared and assigned a value before`

Reviewer 2: I think we should not raise when the variable occurrence at the first instance. In the following example we should raise on VAR1 with value2 only
env:
    - name: VAR1
      value: "value1"
    - name: VAR1 # Noncompliant
      value: "value2"
